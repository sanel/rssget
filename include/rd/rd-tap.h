#ifndef __RD_TAP_H__
#define __RD_TAP_H__

/* TAP - test anything protocol */

void rd_plan(int n);
int  rd_done(void);

void rd_todo(const char *msg);
void rd_todo_end();

void rd_skip(int n, const char *msg);
void rd_skip_end();

int _rd_ok(int expr, const char *msg, const char *path, unsigned int line);
int _rd_is_str(const char *s1, const char *s2, const char *msg, const char *path, unsigned int line);

#define rd_ok(expr, str)        _rd_ok(expr, str, __FILE__, __LINE__)
#define rd_not_ok(expr, str)    _rd_ok(!(expr), str, __FILE__, __LINE__)
#define rd_is_str(s1, s2, str)  _rd_is_str(s1, s2, str, __FILE__, __LINE__)

#endif
