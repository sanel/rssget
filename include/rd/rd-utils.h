#ifndef __RD_UTILS_H__
#define __RD_UTILS_H__

#include <assert.h>

#define RD_INLINE __inline__

#define RD_DEBUG(...)   rd_verbose_printf(RD_STRLOC ": ", __VA_ARGS__)
#define RD_WARNING printf
#define RD_ASSERT  assert

#define _RD_STRLOC_STRINGIFY(arg)          _RD_STRLOC_STRINGIFY_ARG(arg)
#define _RD_STRLOC_STRINGIFY_ARG(content)  #content
#define RD_STRLOC                          __FILE__ ":" _RD_STRLOC_STRINGIFY(__LINE__)

#if defined(__GNUC__) && (__GNUC__ > 2) && defined(__OPTIMIZE__)
	#define _RD_BOOLEAN_EXPR(expr)				\
	__extension__ ({							\
			int _rd_boolean_var_;				\
			if(expr)							\
				_rd_boolean_var_ = 1;			\
			else								\
				_rd_boolean_var_ = 0;			\
			_rd_boolean_var_;                   \
	})

	#define RD_LIKELY(expr)   (__builtin_expect(_RD_BOOLEAN_EXPR(expr), 1))
	#define RD_UNLIKELY(expr) (__builtin_expect(_RD_BOOLEAN_EXPR(expr), 0))
#else
	#define RD_LIKELY(expr)   (expr)
	#define RD_UNLIKELY(expr) (expr)
#endif

#define RD_RETURN_IF_FAIL(expr)											\
	do {																\
		if RD_LIKELY(expr) { }											\
		else {															\
			RD_WARNING(RD_STRLOC ": Condition '%s' failed\n", #expr);	\
			return;														\
		}                                                               \
	} while(0)

#define RD_RETURN_VAL_IF_FAIL(expr, val)								\
	do {																\
		if RD_LIKELY(expr) { }											\
		else {															\
			RD_WARNING(RD_STRLOC ": Condition '%s' failed\n", #expr);	\
			return (val);												\
		}																\
	} while(0)

/** Returns home folder. */
const char *
rd_home_dir(void);

/** Global setting for verbosity. */
void
rd_set_verbose(int v);

/** Return 1 if we should be verbose. */
int
rd_is_verbose(void);

/** Used from RD_DEBUG macro. */
int
rd_verbose_printf(const char *pos, const char *fmt, ...);

/** Concat string using printf form. Returns allocated string that should be free()-ed. */
char *
rd_string_printf(const char *fmt, ...);

#endif
