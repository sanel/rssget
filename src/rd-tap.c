#include <rd/rd-tap.h>
#include <stdio.h>
#include <string.h>

static int texpected;
static int tfailed;
static int tcurr;
static const char *todo_msg;

void rd_plan(int n) {
	texpected = n;
	printf("1..%i\n", n);
}

int rd_done(void) {
	int ret = 0;

	if (texpected > tcurr)  {
		printf("# Looks like you have planned %i but executed %i tests\n", texpected, tcurr);
		ret = 1;
	}

	if (tfailed) {
		printf("# Looks like you have failed %i tests out of %i run and %i planned\n",
			   tfailed, tcurr, texpected);
		ret = 1;
	}

	return ret;
}

int _rd_ok(int expr, const char *name, const char *path, unsigned int line) {
	if (!expr) printf("not ");
	printf("ok %i", ++tcurr);
	if (name) printf(" - %s", name);
	if (todo_msg) printf(" # TODO %s", todo_msg);

	printf("\n");

	if (!expr) {
		printf("#   Failed");
		if (todo_msg) printf("(TODO) ");
		printf("test ");
		if (name) printf("'%s'\n#   ", name);
		printf("at %s line %i.\n", path, line);
		if(!todo_msg)
			tfailed++;
	}

	return expr;
}

int _rd_is_str(const char *s1, const char *s2, const char *msg, const char *path, unsigned int line) {
	return _rd_ok(strcmp(s1, s2) == 0, msg, path, line);
}

void rd_todo(const char *msg) {
	todo_msg = msg;
}

void rd_todo_end(void) {
	todo_msg = 0;
}
