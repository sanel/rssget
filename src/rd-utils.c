#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <rd/rd-utils.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <pwd.h>

const char *
rd_home_dir(void) {
	const char *ret = getenv("HOME");
	if(!ret)
		ret = getpwuid(getuid())->pw_dir;
	return ret;
}
