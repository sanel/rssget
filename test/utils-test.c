#include <rd/rd-tap.h>
#include <rd/rd-utils.h>

int main() {
	rd_plan(2);

	rd_ok(rd_is_verbose() == 0, "default verbosity is 0");
	rd_set_verbose(1);
	rd_ok(rd_is_verbose() == 1, "verbosity is 1 now");
	rd_set_verbose(0);

	return rd_done();
}
