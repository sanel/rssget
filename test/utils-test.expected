1..2
ok 1 - default verbosity is 0
ok 2 - verbosity is 1 now
